inst/include/dust/interface.hpp

```diff
@@ -79,12 +80,21 @@ cpp11::writable::doubles dust_simulate(cpp11::sexp r_steps,
   cpp11::writable::doubles ret(index.size() * data.size() * steps.size());
 
   std::vector<real_t> dat =
-    dust_simulate<T>(steps, data, state, index, n_threads, seed);
+    dust_simulate<T>(steps, data, state, index, n_threads, seed, return_state);
   std::copy(dat.begin(), dat.end(), REAL(ret));
 
   ret.attr("dim") = cpp11::writable::integers({(int)index.size(),
                                                (int)data.size(),
                                                (int)steps.size()});
+
+  if (return_state) {
+    cpp11::writable::doubles r_state_end(state.size());
+    std::copy(state.begin(), state.end(), REAL(r_state_end));
```

Comment:
No way around the two copies due to conflicting std::vector and cpp11 vector types?

Reply:
exactly - we'd have to template much of the underlying functions and make sure we had a proxy class that made the REALSXP behave enough like an iterator - it could be done but not in this PR :)